File: backend/api/schemas/catalog.py
"""
API schemas for the Dataset Catalog endpoints.

IMPORTANT: file_path is intentionally absent from all response schemas.
    Raw filesystem paths must remain backend-only.
    Consumers interact exclusively through catalog_id and metadata.
"""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class RegisterDatasetRequest(BaseModel):
    """Input for POST /catalog/datasets — registers a local file by path."""
    provider_type: str          # "csv" | "parquet"
    file_path: str              # backend-only; not echoed in responses
    display_name: str
    symbol: str
    asset_class: str
    venue: str
    timeframe: str
    dataset_type: str = "ohlcv"
    adjustment_mode: str = "adjusted"
    metadata: Optional[dict[str, Any]] = None


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class CatalogEntryResponse(BaseModel):
    """Single catalog entry — no file_path."""
    catalog_id: str
    provider_type: str
    display_name: str
    dataset_type: str
    asset_class: str
    timeframe: str
    symbol: str
    venue: str
    adjustment_mode: str
    registered_at: datetime
    enabled: bool
    metadata: Optional[dict[str, Any]] = None


class CatalogListResponse(BaseModel):
    entries: list[CatalogEntryResponse]
    count: int


class RegisterDatasetResponse(BaseModel):
    catalog_id: str
    display_name: str
    provider_type: str
    symbol: str
    asset_class: str
    venue: str
    timeframe: str


class CatalogOHLCVCandle(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float


class CatalogOHLCVResponse(BaseModel):
    catalog_id: str
    provider_type: str
    symbol: str
    asset_class: str
    venue: str
    timeframe: str
    candle_count: int
    candles: list[CatalogOHLCVCandle]

File: frontend/src/types/catalog.ts
/**
 * Catalog types — Phase 3P-E.
 *
 * IMPORTANT: file_path does NOT appear in any response type.
 * File path is a registration-time-only input; backend resolves it internally.
 * Frontend durable identity is catalog_id only.
 */

export type CatalogProviderType = 'csv' | 'parquet'
export type CatalogTimeframe    = '1d' | '1w' | '1M' | '1h' | '30m' | '15m' | '5m' | '1m'
export type CatalogAssetClass   = 'equity' | 'etf' | 'crypto' | 'fx' | 'future' | 'index'
export type CatalogAdjustment   = 'adjusted' | 'unadjusted'

// ---------------------------------------------------------------------------
// Request — file_path is accepted for registration only; never echoed back
// ---------------------------------------------------------------------------

export interface RegisterDatasetRequest {
  file_path:       string                // registration-time input; cleared after submit
  display_name:    string
  provider_type:   CatalogProviderType
  symbol:          string
  asset_class:     string
  timeframe:       string
  venue:           string
  dataset_type?:   string                // defaults to "ohlcv" on backend
  adjustment_mode?: CatalogAdjustment
}

// ---------------------------------------------------------------------------
// Responses — no file_path
// ---------------------------------------------------------------------------

export interface CatalogEntry {
  catalog_id:      string
  provider_type:   string
  display_name:    string
  dataset_type:    string
  asset_class:     string
  timeframe:       string
  symbol:          string
  venue:           string
  adjustment_mode: string
  registered_at:   string
  enabled:         boolean
  metadata?:       Record<string, unknown> | null
}

export interface CatalogListResponse {
  entries: CatalogEntry[]
  count:   number
}

export interface RegisterDatasetResponse {
  catalog_id:    string
  display_name:  string
  provider_type: string
  symbol:        string
  asset_class:   string
  venue:         string
  timeframe:     string
}

export interface CatalogOHLCVCandle {
  timestamp: string
  open:      number
  high:      number
  low:       number
  close:     number
  volume:    number
}

export interface CatalogOHLCVResponse {
  catalog_id:    string
  provider_type: string
  symbol:        string
  asset_class:   string
  venue:         string
  timeframe:     string
  candle_count:  number
  candles:       CatalogOHLCVCandle[]
}
